source/Core/Drivers/FS2711.cpp

```diff
@@ -0,0 +1,334 @@
+  uint8_t state1 = ~i2c_read(FS2711_REG_STATE1);
+
+  // state.state = (state1 | ((uint16_t)state0) << 8);
+  state.state = state1;
+
+  if (state1 & 0x1) {
+    state.proto_exists = i2c_read(FS2711_REG_PROTOCOL_EXISTS);
+    state.proto_exists |= i2c_read(FS2711_REG_PROTOCOL_EXISTS + 1) << 8;
+    state.proto_exists |= i2c_read(FS2711_REG_PROTOCOL_EXISTS + 2) << 16;
+
+    state.qc_max_volt = i2c_read(FS2711_PROTOCOL_QC_MAX_VOLT);
+  }
+
+  if (state1 & 0x2) {
+    state.pdo_num = 0;
+    memset(state.pdo_type, 0, 7);
+    memset(state.pdo_min_volt, 0, 7);
+    memset(state.pdo_max_volt, 0, 7);
+    memset(state.pdo_max_curr, 0, 7);
+    osDelay(5000);
+
+    for (i = 0; 7 > i; i++) {
+      pdo_b0 = i2c_read(FS2711_REG_PDO_B0 + i * 4);
+      pdo_b1 = i2c_read(FS2711_REG_PDO_B1 + i * 4);
+      pdo_b2 = i2c_read(FS2711_REG_PDO_B2 + i * 4);
+      pdo_b3 = i2c_read(FS2711_REG_PDO_B3 + i * 4);
+
+      if (pdo_b0) {
+        if (pdo_b3 & 0xC0) {
+          state.pdo_type[i]     = FS2711_PDO_PPS;
+          state.pdo_min_volt[i] = pdo_b1 * 100;
+          state.pdo_max_volt[i] = ((pdo_b2 >> 1) + ((pdo_b3 & 0x1) << 7)) * 100;
+          state.pdo_max_curr[i] = (pdo_b0 & 0x7F) * 50;
+        } else {
+          state.pdo_type[i]     = FS2711_PDO_FIX;
+          state.pdo_min_volt[i] = ((pdo_b1 >> 2) + ((pdo_b2 & 0xF) << 6)) * 50;
+          state.pdo_max_volt[i] = state.pdo_min_volt[i];
+          state.pdo_max_curr[i] = (pdo_b0 + ((pdo_b1 & 0x3) << 8)) * 10;
+        }
+        state.pdo_num++;
+      }
+    }
+  }
+}
+
+void FS2711::probe_pd() {
+  enable_protocol(false);
+  osDelay(500);
+  select_protocol(FS2711_PROTOCOL_PD);
+  enable_protocol(true);
+  state.protocol = FS2711_PROTOCOL_PD;
+
+  uint8_t i      = 0;
+  uint8_t pdo_b0 = 0, pdo_b1 = 0, pdo_b2 = 0, pdo_b3 = 0;
+
+  state.pdo_num = 0;
+  memset(state.pdo_type, 0, 7);
+  memset(state.pdo_min_volt, 0, 7);
+  memset(state.pdo_max_volt, 0, 7);
+  memset(state.pdo_max_curr, 0, 7);
+  osDelay(5000);
+
+  for (i = 0; 7 > i; i++) {
+    pdo_b0 = i2c_read(FS2711_REG_PDO_B0 + i * 4);
+    pdo_b1 = i2c_read(FS2711_REG_PDO_B1 + i * 4);
+    pdo_b2 = i2c_read(FS2711_REG_PDO_B2 + i * 4);
+    pdo_b3 = i2c_read(FS2711_REG_PDO_B3 + i * 4);
+
+    if (pdo_b0) {
+      if (pdo_b3 & 0xC0) {
+        state.pdo_type[i]     = FS2711_PDO_PPS;
+        state.pdo_min_volt[i] = pdo_b1 * 100;
+        state.pdo_max_volt[i] = ((pdo_b2 >> 1) + ((pdo_b3 & 0x1) << 7)) * 100;
+        state.pdo_max_curr[i] = (pdo_b0 & 0x7F) * 50;
+      } else {
+        state.pdo_type[i]     = FS2711_PDO_FIX;
+        state.pdo_min_volt[i] = ((pdo_b1 >> 2) + ((pdo_b2 & 0xF) << 6)) * 50;
+        state.pdo_max_volt[i] = state.pdo_min_volt[i];
+        state.pdo_max_curr[i] = (pdo_b0 + ((pdo_b1 & 0x3) << 8)) * 10;
+      }
+      state.pdo_num++;
+    }
+  }
+}
+
+fs2711_state_t FS2711::get_state() { return state; }
+
+bool FS2711::open_pps(uint8_t PDOID, uint16_t volt, uint16_t max_curr) {
```

Comment:
The paremeters are mixed UPPERCASE and snake_case. Should ideally keep to one or the other?